### DISCUSSION TITLE
TypeError: 'GeometryCollection' object is not subscriptable

### DISCUSSION THREAD

**Discussion by jetsonwork** [obss/sahi#1011] [Q&A] [answered]
Hi I am getting the below error and I cannot complete the slicing, could you please advice how can I fix it?

 TypeError: 'GeometryCollection' object is not subscriptable

**Comment by clairelin23** [comment]
I also get the same problem

**Comment by Preburk** [comment]
Same here, feels like a dependency has been updated and broke it

**Comment by fcakyon** [comment]
Have you tried with `shapely==1.8.5`? @Preburk @clairelin23 @jetsonwork 

Moreover, all continuous integration tests for slicing operations are passing with the `shapely==2.0.4`

**Reply by PolyubinAI** [reply]
Hi, man!

shapely==1.8.5
slice_coco doesn't work with import error: 
can't import MultiPolygon from shapely:

and also shapely==2.0.4 raises error: 
TypeError: 'GeometryCollection' object is not subscriptable

**Reply by Alias-z** [reply]
I have the same problem here

**Reply by PolyubinAI** [reply]
I solve this problem with such case:

In my image, the bbox had a border right at the end of the snapshot. That is, I have a 384*384 snapshot and the bbox ends in a pixel with coordinates (30, 384).

As soon as I changed the end of the box to the value (30, 383.95) in the markup service, the year started working and everything went well

**Reply by Alias-z** [reply]
Hi Thanks. Follow your direction, I found the causation and potential solution:

https://github.com/obss/sahi/blob/main/sahi/utils/coco.py#L226-L240. The polygon must be valid in shapely. for example:

![linearring](https://github.com/obss/sahi/assets/66273343/1623374e-71b6-4ee1-8863-a3dd27b554ff)
![polygon2](https://github.com/obss/sahi/assets/66273343/ea9e1344-35e1-4a34-ad9a-d6709abf7177)
![polygon](https://github.com/obss/sahi/assets/66273343/752b3c06-efca-46d7-b6d6-6f3712282ecc)

(https://shapely.readthedocs.io/en/stable/manual.html)

My solution is to pre-process my annotations before I convert them to COCO:
```
            for obj in data['objects']:
                polygon = Polygon(obj['segmentation'])  # convert segmentation to a Shapely Polygon and check validity
                if not polygon.is_valid:
                    invalid_reason = explain_validity(polygon)  # the reason for invalidity
                    print(f"Polygon from {json_path} is invalid: {invalid_reason}")  # print the reason for invalidity
                    polygon = polygon.buffer(0)  # correcting the polygon if it's invalid
```

I think the developer can do something similar with COCO annotation more efficiently. Otherwise I can find a way to do a PR

**Reply by fcakyon** [reply]
@Alias-z and @PolyubinAI, thanks a lot for explaining the core issue!

@Alias-z if you fix the errors in the tests, we are good to merge your PR and solve this issue 💯

**Comment by fcakyon** [comment, marked as answer]
@jetsonwork @clairelin23 @PolyubinAI @Preburk, this issue is fixed in the latest pr thanks to @Alias-z 🚀 
https://github.com/obss/sahi/pull/1047
